desc/protoparse/linker.go

```diff
@@ -1111,3 +1115,100 @@ func (l *linker) checkExtension(fld *desc.FieldDescriptor, res *parseResult) err
 
 	return nil
 }
+
+func (l *linker) checkJsonNamesInFile(fd *desc.FileDescriptor, res *parseResult) error {
+	for _, md := range fd.GetMessageTypes() {
+		if err := l.checkJsonNamesInMessage(md, res); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (l *linker) checkJsonNamesInMessage(md *desc.MessageDescriptor, res *parseResult) error {
+	if err := checkJsonNames(md, res, false); err != nil {
+		return err
+	}
+	if err := checkJsonNames(md, res, true); err != nil {
+		return err
+	}
+	return nil
+}
+
+func checkJsonNames(md *desc.MessageDescriptor, res *parseResult, useCustom bool) error {
+	type seenName struct {
+		source *dpb.FieldDescriptorProto
+		// field's original JSON nane (which can differ in case from map key)
+		orig string
+		// true if orig is a custom JSON name (vs. the field's default JSON name)
+		custom bool
+	}
+	seen := map[string]seenName{}
+
+	for _, fd := range md.GetFields() {
+		scope := "field " + md.GetName() + "." + fd.GetName()
+		defaultName := internal.JsonName(fd.GetName())
+		name := defaultName
+		custom := false
+		if useCustom {
+			n := fd.GetJSONName()
+			if n != defaultName || hasCustomJsonName(res, fd) {
+				name = n
+				custom = true
+			}
+		}
+		lcaseName := strings.ToLower(name)
+		if existing, ok := seen[lcaseName]; ok {
+			// When useCustom is true, we'll only report an issue when a conflict is
+			// due to a custom name. That way, we don't double report conflicts on
+			// non-custom names.
+			if !useCustom || custom || existing.custom {
+				fldNode := res.getFieldNode(fd.AsFieldDescriptorProto())
+				customStr, srcCustomStr := "custom", "custom"
+				if !custom {
+					customStr = "default"
+				}
+				if !existing.custom {
+					srcCustomStr = "default"
+				}
+				otherName := ""
+				if name != existing.orig {
+					otherName = fmt.Sprintf(" %q", existing.orig)
+				}
+				conflictErr := errorWithPos(fldNode.Start(), "%s: %s JSON name %q conflicts with %s JSON name%s of field %s, defined at %v",
```

Comment:
```suggestion
				conflictErr := errorWithPos(fldNode.Start(), "%s: %s JSON name %q conflicts with %s JSON name %s of field %s, defined at %v",
```

Reply:
I would have thought we'd need a space here, but is there always a space in the value we're including here? By the error messages included in your test, everything looks correct, so I'm a little surprised.

This would be the `existing.source.GetName()` value.

Reply:
We only print a name here when the two names that conflict differ (which can happen if they vary only by case). So if the previously defined JSON name is the same as the one causing the conflict, we omit it entirely. See how the name var is actually constructed above: if it's not empty, it includes a leading space.

Reply:
Ah ha! Thanks for the clarification - makes sense.

Reply:
FWIW, this is actually for the `origValue` value. The `existing.source.GetName()` is the next `%s` in the format string.